test/test_source_hdf5.py

```diff
@@ -0,0 +1,177 @@
+import pathlib
+import random
+import tempfile
+
+import numpy
+
+import large_image
+
+possible_axes = {
+    'x': [1, 10],
+    'y': [1, 10],
+    'c': [1, 100],
+    'z': [1, 100],
+    't': [1, 100],
+    'p': [1, 100],
+    'q': [1, 100],
+    's': [3, 3],
+}
+
+include_axes = {
+    'c': False,
+    'z': False,
+    't': False,
+    'p': False,
+    'q': False,
+}
+
+possible_data_ranges = [
+    [0, 1, float],
+    [0, 2**8, numpy.uint8],
+    [0, 2**16, numpy.uint16],
+    [0, 2**32, numpy.uint32],
+    [-2**7, 2**7, numpy.int8],
+    [-2**15, 2**15, numpy.int16],
+    [-2**31, 2**31, numpy.int32],
+    [-1, 1, float]
+]
+
+max_tile_size = 100
+tile_overlap_ratio = 0.5
+
+
+# https://stackoverflow.com/questions/18915378/rounding-to-significant-figures-in-numpy
+def signif(x):
+    if x == 0:
+        return 0
+    return round(x, -2)
+
+
+def get_dims(x, y, s, max=False):
+    tile_shape = [x, y]
+    for axis_name, include in include_axes.items():
+        if include:
+            axis_min_max = possible_axes[axis_name]
+            if max:
+                tile_shape.append(axis_min_max[1])
+            else:
+                tile_shape.append(random.randint(*axis_min_max))
+    # s is last axis
+    tile_shape.append(s)
+    return tile_shape
+
+
+def random_tile(data_range):
+    tile_shape = get_dims(
+        random.randint(1, max_tile_size),
+        random.randint(1, max_tile_size),
+        random.randint(*possible_axes['s']),
+        include_axes
+    )
+    tile = numpy.random.rand(*tile_shape)
+    tile *= (data_range[1] - data_range[0])
+    tile += data_range[0]
+    tile = tile.astype(data_range[2])  # apply dtype
+    mask = numpy.random.randint(2, size=tile_shape[:-1])
+    return (tile, mask)
+
+
+def frame_with_zeros(data, desired_size, start_location=None):
+    if len(desired_size) == 0:
+        return data
+    if not start_location or len(start_location) == 0:
+        start_location = [0]
+    framed = [
+        frame_with_zeros(
+            data[x - start_location[0]],
+            desired_size[1:],
+            start_location=start_location[1:]
+        )
+        if (  # frame with zeros if x>=start and x<start+length
+            x >= start_location[0] and
+            x < data.shape[0] + start_location[0]
+        )  # fill with zeros otherwise
+        else numpy.zeros(desired_size[1:])
+        for x in range(desired_size[0])
+    ]
+    return numpy.array(framed)
+
+
+def testImageGeneration():
```

Comment:
One last thing before we merge this: I'd like to use `@pytest.mark.parametrize` to try multiple conditions.  I think initially we can just want through all of the data ranges:
```
@pytest.mark.parametrize('data_range', possible_data_ranges)
def testImageGeneration(data_range):
```
and, so that the command line version works, we would change the main run command to `testImageGeneration(None)`, and in the `testImageGeneration` funcion check if `data_range` is None, and, if so, set it to the random choice.

Reply:
Changed in [71cca3d](https://github.com/girder/large_image/pull/1011/commits/71cca3dac7db7800207d7553fb04230fcd628e1b)